models/classes/resources/Service/ClassMover.php

```diff
@@ -0,0 +1,178 @@
+
+
+            if (isset($this->permissionCopier) && $command->useDestinationAcl()) {
+                $this->changePermissions($to, $from);
+            }
+        }
+
+        return new ResourceTransferResult($from->getUri());
+    }
+
+    private function changePermissions(
+        core_kernel_classes_Class $source,
+        core_kernel_classes_Class $destination
+    ): void {
+        $this->permissionCopier->copy($source, $destination);
+
+        foreach ($destination->getInstances() as $instance) {
+            $this->permissionCopier->copy($source, $instance);
+        }
+
+        foreach ($destination->getSubClasses() as $subClass) {
+            $this->changePermissions($destination, $subClass);
+        }
+    }
+
+    private function assertIsNotRootClass(core_kernel_classes_Class $class): self
+    {
+        if ($this->rootClassSpecification->isSatisfiedBy($class)) {
+            throw new InvalidArgumentException(sprintf('Root class "%s" cannot be moved', $class->getUri()));
+        }
+
+        return $this;
+    }
+
+    private function assertInSameRootClass(
+        core_kernel_classes_Class $source,
+        core_kernel_classes_Class $destionation
+    ): self {
+        foreach ($this->rootClassesListService->list() as $rootClass) {
+            if (
+                ($source->equals($rootClass) || $source->isSubClassOf($rootClass))
+                && !$destionation->equals($rootClass)
+                && !$destionation->isSubClassOf($rootClass)
+            ) {
+                throw new InvalidArgumentException(
+                    sprintf(
+                        'Selected class (%s) and destination class (%s) must be in the same root class (%s).',
+                        $source->getUri(),
+                        $destionation->getUri(),
+                        $rootClass->getUri()
+                    )
+                );
+            }
+        }
+
+        return $this;
+    }
+
+    private function assertIsNotSameClass(
+        core_kernel_classes_Class $source,
+        core_kernel_classes_Class $destination
+    ): self {
+        if ($source->equals($destination)) {
+            throw new InvalidArgumentException(
+                sprintf(
+                    'Selected class (%s) and destination class (%s) cannot be the same class.',
+                    $source->getUri(),
+                    $destination->getUri()
+                )
+            );
+        }
+
+        return $this;
+    }
+
+    private function assertIsNotSubclass(
+        core_kernel_classes_Class $source,
+        core_kernel_classes_Class $destination
+    ): void {
+        if ($destination->isSubClassOf($source)) {
+            throw new InvalidArgumentException(
+                sprintf(
+                    'The destination class (%s) cannot be a subclass of the selected class (%s).',
+                    $destination->getUri(),
+                    $source->getUri()
+                )
+            );
+        }
```

Comment:
nitpick: I'd avoid returning `$this` from an assertion and write the chain of assertions as separate statements instead (i.e. make all of them them a sort-of command checking a condition and not _producing_ anything):

```php
$this->assertIsNotRootClass($from)
$this->assertInSameRootClass($from, $to)
$this->assertIsNotSameClass($from, $to)
$this->assertIsNotSubclass($from, $to);
```

But if we go that way, I think we should also return `$this` from this one for consistency with the others even if that is not technically needed for the order they are called _right now_ :slightly_smiling_face: 

```suggestion
    private function assertIsNotSubclass(
        core_kernel_classes_Class $source,
        core_kernel_classes_Class $destination
    ): self {
        if ($destination->isSubClassOf($source)) {
            throw new InvalidArgumentException(
                sprintf(
                    'The destination class (%s) cannot be a subclass of the selected class (%s).',
                    $destination->getUri(),
                    $source->getUri()
                )
            );
        }
        
        return $this;
```

Reply:
Done